{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import regex\n",
    "import re\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import Bio\n",
    "from Bio import motifs\n",
    "from Bio import pairwise2 \n",
    "from Bio.pairwise2 import format_alignment\n",
    "import collections\n",
    "import itertools\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Install package inside Jupter Notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting openpyxl\n",
      "  Downloading openpyxl-3.1.1-py2.py3-none-any.whl (249 kB)\n",
      "     -------------------------------------- 249.8/249.8 kB 3.9 MB/s eta 0:00:00\n",
      "Collecting et-xmlfile\n",
      "  Downloading et_xmlfile-1.1.0-py3-none-any.whl (4.7 kB)\n",
      "Installing collected packages: et-xmlfile, openpyxl\n",
      "Successfully installed et-xmlfile-1.1.0 openpyxl-3.1.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "[notice] A new release of pip available: 22.2.2 -> 23.0.1\n",
      "[notice] To update, run: C:\\Users\\shangl02\\AppData\\Local\\Microsoft\\WindowsApps\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\python.exe -m pip install --upgrade pip\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!{sys.executable} -m pip install openpyxl\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AS15-01_S1.fastq', 'AS15-02_S2.fastq', 'AS15-03_S3.fastq', 'AS15-04_S4.fastq', 'AS15-05_S5.fastq', 'AS15-06_S6.fastq', 'AS15-07_S7.fastq', 'AS15-08_S8.fastq', 'AS15-09_S9.fastq', 'AS15-10_S10.fastq', 'AS15-11_S11.fastq', 'AS15-12_S12.fastq', 'AS15-13_S13.fastq', 'AS15-14_S14.fastq', 'AS15-15_S15.fastq', 'AS15-16_S16.fastq', 'AS15-17_S17.fastq', 'AS15-18_S18.fastq', 'AS15-19_S19.fastq', 'AS15-20_S20.fastq', 'AS15-21_S21.fastq', 'AS15-22_S22.fastq', 'AS15-23_S23.fastq', 'AS15-24_S24.fastq', 'AS15-25_S25.fastq', 'AS15-26_S26.fastq', 'AS15-27_S27.fastq', 'AS15-28_S28.fastq', 'AS15-29_S29.fastq', 'AS15-30_S30.fastq', 'AS15-31_S31.fastq', 'AS15-32_S32.fastq', 'AS15-33_S33.fastq', 'AS15-34_S34.fastq', 'AS15-35_S35.fastq', 'AS15-36_S36.fastq', 'AS15-37_S37.fastq', 'AS15-38_S38.fastq', 'AS15-39_S39.fastq', 'AS15-40_S40.fastq', 'AS15-41_S41.fastq', 'AS15-42_S42.fastq', 'AS15-43_S43.fastq', 'AS15-44_S44.fastq', 'AS15-45_S45.fastq', 'AS15-46_S46.fastq', 'AS15-47_S47.fastq', 'AS15-48_S48.fastq', 'AS15-49_S49.fastq', 'AS15-50_S50.fastq', 'AS15-51_S51.fastq', 'AS15-52_S52.fastq', 'AS15-53_S53.fastq', 'AS15-54_S54.fastq', 'AS15-55_S55.fastq', 'AS15-56_S56.fastq', 'AS15-57_S57.fastq', 'AS15-58_S58.fastq', 'AS15-59_S59.fastq', 'AS15-60_S60.fastq', 'AS15-61_S61.fastq', 'AS15-62_S62.fastq', 'AS15-63_S63.fastq', 'AS15-64_S64.fastq', 'AS15-65_S65.fastq', 'AS15-66_S66.fastq', 'AS15-67_S67.fastq', 'AS15-68_S68.fastq', 'AS15-69_S69.fastq', 'AS15-70_S70.fastq', 'AS15-71_S71.fastq', 'AS15-72_S72.fastq', 'AS15-73_S73.fastq', 'AS15-74_S74.fastq', 'AS15-75_S75.fastq', 'AS15-76_S76.fastq', 'AS15-77_S77.fastq', 'AS15-78_S78.fastq', 'AS15-79_S79.fastq', 'AS15-80_S80.fastq', 'AS15-81_S81.fastq', 'AS15-82_S82.fastq', 'AS15-83_S83.fastq', 'AS15-84_S84.fastq', 'AS15-85_S85.fastq', 'AS15-86_S86.fastq', 'AS15-87_S87.fastq', 'AS15-88_S88.fastq', 'AS15-89_S89.fastq', 'AS15-90_S90.fastq', 'AS15-91_S91.fastq', 'AS15-92_S92.fastq']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# set up working directory\n",
    "os.chdir(r'X:\\users\\shangl02\\Test\\ampliCan.test2\\AS15\\fastq.merge')\n",
    "directory = os.getcwd()\n",
    "#filenames = os.listdir(directory, '*.fastq')\n",
    "filenames=[]\n",
    "for file in os.listdir(directory):\n",
    "    if file.endswith(\".fastq\"):\n",
    "        filenames.append(file)\n",
    "print(filenames)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Sample Number    Sample Name  Experiment #  \\\n",
      "0        AS15_01  BCL_GFP/500_1             1   \n",
      "1        AS15_02  BCL_GFP/500_2             1   \n",
      "2        AS15_03  BCL_GFP/500_3             1   \n",
      "3        AS15_04  BCL_IDT/500_1             1   \n",
      "4        AS15_05  BCL_IDT/500_2             1   \n",
      "..           ...            ...           ...   \n",
      "87       AS15_88     HBG_sg44_3             2   \n",
      "88       AS15_89     HBG_sg45_3             2   \n",
      "89       AS15_90     HBG_sg46_3             2   \n",
      "90       AS15_91     HBG_sg50_3             2   \n",
      "91       AS15_92     HBG_sg30_3             2   \n",
      "\n",
      "                                        Full Amplicon  \\\n",
      "0   TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCAGTGCAAAGTC...   \n",
      "1   TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCAGTGCAAAGTC...   \n",
      "2   TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCAGTGCAAAGTC...   \n",
      "3   TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCAGTGCAAAGTC...   \n",
      "4   TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCAGTGCAAAGTC...   \n",
      "..                                                ...   \n",
      "87  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...   \n",
      "88  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...   \n",
      "89  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...   \n",
      "90  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...   \n",
      "91  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...   \n",
      "\n",
      "                   Expected Sequence (if base editor)        Guide Sequence  \\\n",
      "0                                                 NaN  CTAACAGTTGCTTTTATCAC   \n",
      "1                                                 NaN  CTAACAGTTGCTTTTATCAC   \n",
      "2                                                 NaN  CTAACAGTTGCTTTTATCAC   \n",
      "3                                                 NaN  CTAACAGTTGCTTTTATCAC   \n",
      "4                                                 NaN  CTAACAGTTGCTTTTATCAC   \n",
      "..                                                ...                   ...   \n",
      "87  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...  AGCCAGGGACCGTTTCAGAC   \n",
      "88  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...  GTTTAGCCAGGGACCGTTTC   \n",
      "89  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...  AAGGCAAGGCTGGCCAACCC   \n",
      "90  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...  AGATATTTGCATTGAGATAG   \n",
      "91  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...  CATTGAGATAGTGTGGGGAA   \n",
      "\n",
      "          Fwd primer bind  SNP_position         Rev primer bind  \\\n",
      "0   CCAGTGCAAAGTCCATACAGG           NaN   GTCTGCCAGTCCTCTTCTACC   \n",
      "1   CCAGTGCAAAGTCCATACAGG           NaN   GTCTGCCAGTCCTCTTCTACC   \n",
      "2   CCAGTGCAAAGTCCATACAGG           NaN   GTCTGCCAGTCCTCTTCTACC   \n",
      "3   CCAGTGCAAAGTCCATACAGG           NaN   GTCTGCCAGTCCTCTTCTACC   \n",
      "4   CCAGTGCAAAGTCCATACAGG           NaN   GTCTGCCAGTCCTCTTCTACC   \n",
      "..                    ...           ...                     ...   \n",
      "87   GAATCGGAACAAGGCAAAGG           NaN  CTTTTATTCTTCATCCCTAGCC   \n",
      "88   GAATCGGAACAAGGCAAAGG           NaN  CTTTTATTCTTCATCCCTAGCC   \n",
      "89   GAATCGGAACAAGGCAAAGG           NaN  CTTTTATTCTTCATCCCTAGCC   \n",
      "90   GAATCGGAACAAGGCAAAGG           NaN  CTTTTATTCTTCATCCCTAGCC   \n",
      "91   GAATCGGAACAAGGCAAAGG           NaN  CTTTTATTCTTCATCCCTAGCC   \n",
      "\n",
      "                                     RC_Full Amplicon     RC_Guide Sequence  \\\n",
      "0   GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTCTGCCAGTCC...  GTGATAAAAGCAACTGTTAG   \n",
      "1   GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTCTGCCAGTCC...  GTGATAAAAGCAACTGTTAG   \n",
      "2   GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTCTGCCAGTCC...  GTGATAAAAGCAACTGTTAG   \n",
      "3   GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTCTGCCAGTCC...  GTGATAAAAGCAACTGTTAG   \n",
      "4   GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTCTGCCAGTCC...  GTGATAAAAGCAACTGTTAG   \n",
      "..                                                ...                   ...   \n",
      "87  GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTTTTATTCTTC...  GTCTGAAACGGTCCCTGGCT   \n",
      "88  GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTTTTATTCTTC...  GAAACGGTCCCTGGCTAAAC   \n",
      "89  GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTTTTATTCTTC...  GGGTTGGCCAGCCTTGCCTT   \n",
      "90  GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTTTTATTCTTC...  CTATCTCAATGCAAATATCT   \n",
      "91  GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTTTTATTCTTC...  TTCCCCACACTATCTCAATG   \n",
      "\n",
      "   guide_in_sense  \n",
      "0            True  \n",
      "1            True  \n",
      "2            True  \n",
      "3            True  \n",
      "4            True  \n",
      "..            ...  \n",
      "87          False  \n",
      "88          False  \n",
      "89          False  \n",
      "90          False  \n",
      "91          False  \n",
      "\n",
      "[92 rows x 12 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# generate metadata dataframe from submitted sequence information sheet including ‘Sample #Number’, ‘Guide Sequence’ and ‘Full Amplicon’. Add two columns indicating reverse complement #sequences\n",
    "meta = pd.read_excel(r'X:\\users\\shangl02\\Test\\ampliCan.test2\\AS15\\AS15_submissionform for Eva.xlsx')\n",
    "meta['RC_Full Amplicon'] = [Bio.Seq.reverse_complement(seq) for seq in meta['Full Amplicon']]\n",
    "meta['RC_Guide Sequence'] = [Bio.Seq.reverse_complement(seq) if isinstance(seq, str) else seq for seq in meta['Guide Sequence']]\n",
    "\n",
    "#Add a column to the metadata dataframe indicating the orientation of guide\n",
    "guide_in_sense = []\n",
    "for index, row in meta.iterrows():\n",
    "    if isinstance(row['Guide Sequence'], float):\n",
    "        check = 'NA'\n",
    "    elif row['Guide Sequence'] in row['Full Amplicon']:\n",
    "        check = True\n",
    "    elif row['Guide Sequence'] in row['RC_Full Amplicon']:\n",
    "        check = False\n",
    "    guide_in_sense.append(check)\n",
    "meta['guide_in_sense'] = guide_in_sense\n",
    "\n",
    "print(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add two columns indicating the 10bp upstream and downstream of protospacer, respectively\n",
    "flank1 = {}\n",
    "flank2 = {}\n",
    "for index, row in meta.iterrows():\n",
    "    if isinstance(row['Guide Sequence'], float):\n",
    "        continue\n",
    "    if row['Guide Sequence'] in row['Full Amplicon']:\n",
    "        flank1[index] = row['Full Amplicon'].split(row['Guide Sequence'])[0][-10:]\n",
    "        flank2[index] = row['Full Amplicon'].split(row['Guide Sequence'])[1][:10]       \n",
    "    elif row['Guide Sequence'] in row['RC_Full Amplicon']:\n",
    "        flank1[index] = row['Full Amplicon'].split(row['RC_Guide Sequence'])[0][-10:]\n",
    "        flank2[index] = row['Full Amplicon'].split(row['RC_Guide Sequence'])[1][:10]\n",
    "flank_df = pd.DataFrame([flank1, flank2], index = ['flank1', 'flank2']).T\n",
    "meta1 = meta.merge(flank_df, left_index = True, right_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Main base calling function\n",
    "def basecall(directory, filename, guide, flank1, flank2, guide_in_sense):\n",
    "    seqs = {}\n",
    "    seq_parse = list(SeqIO.parse(os.path.join(directory, filename), 'fastq'))\n",
    "    for record in seq_parse:\n",
    "        recordqual = [x > 31 for x in record.letter_annotations['phred_quality']]\n",
    "        if float(sum(recordqual)) / float(len(recordqual)) > 0.5:\n",
    "            recordseq = ''.join([y if x else 'N' for (x,y) in zip(recordqual, record.seq)])\n",
    "            split1 = recordseq.split(flank1)\n",
    "            if len(split1) == 2:\n",
    "                split2 = split1[1].split(flank2)[0]\n",
    "                if len(split2) == 20:\n",
    "                    if guide_in_sense == False:\n",
    "                        seqs[record.id] = Bio.Seq.reverse_complement(split2)\n",
    "                    else:\n",
    "                        seqs[record.id] = split2\n",
    "    frame = pd.DataFrame(list(seqs.items()), columns = ['ID', 'spacer'])\n",
    "    frame1 = frame[['N' not in x for x in frame.spacer]]\n",
    "    Motif = motifs.create(frame1.spacer.values)\n",
    "    raw = pd.DataFrame(Motif.counts, index = [str(i+1) for i in range(20)]).transpose()\n",
    "    normalized = raw / raw.sum(axis = 0) * 100\n",
    "    normalized = normalized.round(2)\n",
    "    guide_split = [*guide]\n",
    "    per_base = pd.concat([pd.DataFrame(guide_split, index = [str(i+1) for i in range(20)]).transpose(), normalized])\n",
    "    counts = frame1.groupby('spacer').count().sort_values('ID', ascending = False)\n",
    "    counts['percentage'] = counts['ID'] / counts['ID'].sum() * 100\n",
    "    return [per_base, counts]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Sample Number    Sample Name  Experiment #  \\\n",
      "0        AS15_01  BCL_GFP/500_1             1   \n",
      "1        AS15_02  BCL_GFP/500_2             1   \n",
      "2        AS15_03  BCL_GFP/500_3             1   \n",
      "3        AS15_04  BCL_IDT/500_1             1   \n",
      "4        AS15_05  BCL_IDT/500_2             1   \n",
      "..           ...            ...           ...   \n",
      "87       AS15_88     HBG_sg44_3             2   \n",
      "88       AS15_89     HBG_sg45_3             2   \n",
      "89       AS15_90     HBG_sg46_3             2   \n",
      "90       AS15_91     HBG_sg50_3             2   \n",
      "91       AS15_92     HBG_sg30_3             2   \n",
      "\n",
      "                                        Full Amplicon  \\\n",
      "0   TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCAGTGCAAAGTC...   \n",
      "1   TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCAGTGCAAAGTC...   \n",
      "2   TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCAGTGCAAAGTC...   \n",
      "3   TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCAGTGCAAAGTC...   \n",
      "4   TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCAGTGCAAAGTC...   \n",
      "..                                                ...   \n",
      "87  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...   \n",
      "88  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...   \n",
      "89  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...   \n",
      "90  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...   \n",
      "91  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...   \n",
      "\n",
      "                   Expected Sequence (if base editor)        Guide Sequence  \\\n",
      "0                                                 NaN  CTAACAGTTGCTTTTATCAC   \n",
      "1                                                 NaN  CTAACAGTTGCTTTTATCAC   \n",
      "2                                                 NaN  CTAACAGTTGCTTTTATCAC   \n",
      "3                                                 NaN  CTAACAGTTGCTTTTATCAC   \n",
      "4                                                 NaN  CTAACAGTTGCTTTTATCAC   \n",
      "..                                                ...                   ...   \n",
      "87  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...  AGCCAGGGACCGTTTCAGAC   \n",
      "88  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...  GTTTAGCCAGGGACCGTTTC   \n",
      "89  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...  AAGGCAAGGCTGGCCAACCC   \n",
      "90  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...  AGATATTTGCATTGAGATAG   \n",
      "91  TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGGAATCGGAACAAG...  CATTGAGATAGTGTGGGGAA   \n",
      "\n",
      "          Fwd primer bind  SNP_position         Rev primer bind  \\\n",
      "0   CCAGTGCAAAGTCCATACAGG           NaN   GTCTGCCAGTCCTCTTCTACC   \n",
      "1   CCAGTGCAAAGTCCATACAGG           NaN   GTCTGCCAGTCCTCTTCTACC   \n",
      "2   CCAGTGCAAAGTCCATACAGG           NaN   GTCTGCCAGTCCTCTTCTACC   \n",
      "3   CCAGTGCAAAGTCCATACAGG           NaN   GTCTGCCAGTCCTCTTCTACC   \n",
      "4   CCAGTGCAAAGTCCATACAGG           NaN   GTCTGCCAGTCCTCTTCTACC   \n",
      "..                    ...           ...                     ...   \n",
      "87   GAATCGGAACAAGGCAAAGG           NaN  CTTTTATTCTTCATCCCTAGCC   \n",
      "88   GAATCGGAACAAGGCAAAGG           NaN  CTTTTATTCTTCATCCCTAGCC   \n",
      "89   GAATCGGAACAAGGCAAAGG           NaN  CTTTTATTCTTCATCCCTAGCC   \n",
      "90   GAATCGGAACAAGGCAAAGG           NaN  CTTTTATTCTTCATCCCTAGCC   \n",
      "91   GAATCGGAACAAGGCAAAGG           NaN  CTTTTATTCTTCATCCCTAGCC   \n",
      "\n",
      "                                     RC_Full Amplicon     RC_Guide Sequence  \\\n",
      "0   GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTCTGCCAGTCC...  GTGATAAAAGCAACTGTTAG   \n",
      "1   GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTCTGCCAGTCC...  GTGATAAAAGCAACTGTTAG   \n",
      "2   GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTCTGCCAGTCC...  GTGATAAAAGCAACTGTTAG   \n",
      "3   GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTCTGCCAGTCC...  GTGATAAAAGCAACTGTTAG   \n",
      "4   GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGGTCTGCCAGTCC...  GTGATAAAAGCAACTGTTAG   \n",
      "..                                                ...                   ...   \n",
      "87  GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTTTTATTCTTC...  GTCTGAAACGGTCCCTGGCT   \n",
      "88  GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTTTTATTCTTC...  GAAACGGTCCCTGGCTAAAC   \n",
      "89  GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTTTTATTCTTC...  GGGTTGGCCAGCCTTGCCTT   \n",
      "90  GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTTTTATTCTTC...  CTATCTCAATGCAAATATCT   \n",
      "91  GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGCTTTTATTCTTC...  TTCCCCACACTATCTCAATG   \n",
      "\n",
      "   guide_in_sense      flank1      flank2  \n",
      "0            True  CTAGTGCAAG  AGGCTCCAGG  \n",
      "1            True  CTAGTGCAAG  AGGCTCCAGG  \n",
      "2            True  CTAGTGCAAG  AGGCTCCAGG  \n",
      "3            True  CTAGTGCAAG  AGGCTCCAGG  \n",
      "4            True  CTAGTGCAAG  AGGCTCCAGG  \n",
      "..            ...         ...         ...  \n",
      "87          False  GCAAATATCT  AAACTCCACC  \n",
      "88          False  ATATCTGTCT  TCCACCCATG  \n",
      "89          False  CTCCACCCAT  GACCAATAGC  \n",
      "90          False  CTTCCCCACA  GTCTGAAACG  \n",
      "91          False  TGGGGGCCCC  CAAATATCTG  \n",
      "\n",
      "[90 rows x 14 columns]\n"
     ]
    }
   ],
   "source": [
    "print(meta1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AS15-01_S1.fastq\n",
      "AS15-02_S2.fastq\n",
      "AS15-03_S3.fastq\n",
      "AS15-04_S4.fastq\n",
      "AS15-05_S5.fastq\n",
      "AS15-06_S6.fastq\n",
      "AS15-07_S7.fastq\n",
      "AS15-08_S8.fastq\n",
      "AS15-09_S9.fastq\n",
      "AS15-10_S10.fastq\n",
      "AS15-11_S11.fastq\n",
      "AS15-12_S12.fastq\n",
      "AS15-13_S13.fastq\n",
      "AS15-14_S14.fastq\n",
      "AS15-15_S15.fastq\n",
      "AS15-16_S16.fastq\n",
      "AS15-17_S17.fastq\n",
      "AS15-18_S18.fastq\n",
      "AS15-19_S19.fastq\n",
      "AS15-20_S20.fastq\n",
      "AS15-21_S21.fastq\n",
      "AS15-22_S22.fastq\n",
      "AS15-23_S23.fastq\n",
      "AS15-24_S24.fastq\n",
      "AS15-25_S25.fastq\n",
      "AS15-26_S26.fastq\n",
      "AS15-27_S27.fastq\n",
      "AS15-28_S28.fastq\n",
      "AS15-29_S29.fastq\n",
      "AS15-30_S30.fastq\n",
      "AS15-31_S31.fastq\n",
      "AS15-32_S32.fastq\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'spacer'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[47], line 8\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[39mif\u001b[39;00m file\u001b[39m.\u001b[39mreplace(\u001b[39m'\u001b[39m\u001b[39m-\u001b[39m\u001b[39m'\u001b[39m,\u001b[39m'\u001b[39m\u001b[39m_\u001b[39m\u001b[39m'\u001b[39m)\u001b[39m.\u001b[39mstartswith(row[\u001b[39m'\u001b[39m\u001b[39mSample Number\u001b[39m\u001b[39m'\u001b[39m]):\n\u001b[0;32m      7\u001b[0m     \u001b[39mprint\u001b[39m(file)\n\u001b[1;32m----> 8\u001b[0m     res[file] \u001b[39m=\u001b[39m basecall(directory, file, row[\u001b[39m'\u001b[39;49m\u001b[39mGuide Sequence\u001b[39;49m\u001b[39m'\u001b[39;49m], \\\n\u001b[0;32m      9\u001b[0m                          row[\u001b[39m'\u001b[39;49m\u001b[39mflank1\u001b[39;49m\u001b[39m'\u001b[39;49m], row[\u001b[39m'\u001b[39;49m\u001b[39mflank2\u001b[39;49m\u001b[39m'\u001b[39;49m], row[\u001b[39m'\u001b[39;49m\u001b[39mguide_in_sense\u001b[39;49m\u001b[39m'\u001b[39;49m])\n",
      "Cell \u001b[1;32mIn[45], line 19\u001b[0m, in \u001b[0;36mbasecall\u001b[1;34m(directory, filename, guide, flank1, flank2, guide_in_sense)\u001b[0m\n\u001b[0;32m     17\u001b[0m frame \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(\u001b[39mlist\u001b[39m(seqs\u001b[39m.\u001b[39mitems()), columns \u001b[39m=\u001b[39m [\u001b[39m'\u001b[39m\u001b[39mID\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mspacer\u001b[39m\u001b[39m'\u001b[39m])\n\u001b[0;32m     18\u001b[0m frame1 \u001b[39m=\u001b[39m frame[[\u001b[39m'\u001b[39m\u001b[39mN\u001b[39m\u001b[39m'\u001b[39m \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m x \u001b[39mfor\u001b[39;00m x \u001b[39min\u001b[39;00m frame\u001b[39m.\u001b[39mspacer]]\n\u001b[1;32m---> 19\u001b[0m Motif \u001b[39m=\u001b[39m motifs\u001b[39m.\u001b[39mcreate(frame1\u001b[39m.\u001b[39;49mspacer\u001b[39m.\u001b[39mvalues)\n\u001b[0;32m     20\u001b[0m raw \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mDataFrame(Motif\u001b[39m.\u001b[39mcounts, index \u001b[39m=\u001b[39m [\u001b[39mstr\u001b[39m(i\u001b[39m+\u001b[39m\u001b[39m1\u001b[39m) \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(\u001b[39m20\u001b[39m)])\u001b[39m.\u001b[39mtranspose()\n\u001b[0;32m     21\u001b[0m normalized \u001b[39m=\u001b[39m raw \u001b[39m/\u001b[39m raw\u001b[39m.\u001b[39msum(axis \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m) \u001b[39m*\u001b[39m \u001b[39m100\u001b[39m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\pandas\\core\\generic.py:5575\u001b[0m, in \u001b[0;36mNDFrame.__getattr__\u001b[1;34m(self, name)\u001b[0m\n\u001b[0;32m   5568\u001b[0m \u001b[39mif\u001b[39;00m (\n\u001b[0;32m   5569\u001b[0m     name \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_internal_names_set\n\u001b[0;32m   5570\u001b[0m     \u001b[39mand\u001b[39;00m name \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_metadata\n\u001b[0;32m   5571\u001b[0m     \u001b[39mand\u001b[39;00m name \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_accessors\n\u001b[0;32m   5572\u001b[0m     \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_info_axis\u001b[39m.\u001b[39m_can_hold_identifiers_and_holds_name(name)\n\u001b[0;32m   5573\u001b[0m ):\n\u001b[0;32m   5574\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m[name]\n\u001b[1;32m-> 5575\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mobject\u001b[39;49m\u001b[39m.\u001b[39;49m\u001b[39m__getattribute__\u001b[39;49m(\u001b[39mself\u001b[39;49m, name)\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'spacer'"
     ]
    }
   ],
   "source": [
    "#Run every fastq file in working directory\n",
    "res = {}\n",
    "for index,row in meta1.iterrows():\n",
    "    for file in filenames:\n",
    "        #if row['Sample Number'] in file:\n",
    "        if file.replace('-','_').startswith(row['Sample Number']):\n",
    "            print(file)\n",
    "            res[file] = basecall(directory, file, row['Guide Sequence'], \\\n",
    "                                 row['flank1'], row['flank2'], row['guide_in_sense'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#there are two outputs from scripts implemented above\n",
    "# 1. per base percentage of each nucleotide\n",
    " \n",
    "\n",
    "# 2. Counts and percentage of each sequence pattern\n",
    " \n",
    "\n",
    "#Get a single value of base editing efficiency from all samples\n",
    "be_eff_all = {}\n",
    "for key, value in res.items():\n",
    "    sample = key[:7]\n",
    "    guide = meta1.loc[meta1['Sample Number'] == sample, 'Guide Sequence'].tolist()[0]\n",
    "    df = value[1][value[1].index != guide]\n",
    "    eff = df['percentage'][0:5].sum()\n",
    "    be_eff_all[key[:7]] = eff\n",
    "be_eff_df = pd.DataFrame(be_eff_all.items(), columns = ['sample', 'base_editing_efficiency'])\n",
    "#the output"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "9ee08997ade7e9cc626f2a6fc3903e4d8ebc0d3b740e112f7f532dcd60319aea"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
